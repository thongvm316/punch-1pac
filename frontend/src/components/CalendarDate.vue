<template>
  <div class="calendar-date current-month" :class="{'disabled': isWeekend}">
    <button class="date-item" :class="{ 'date-today': localAttendance.day === today.format('YYYY-MM-DD') }">
      {{ localAttendance.day.split('-')[2] }}
    </button>
    <div class="calendar-events">
      <span class="calendar-event bg-success text-success text-center" v-if="localAttendance.attending_status === 'attend_ok'">
        {{ $t(`meta.attendance_statuses.${localAttendance.attending_status}`) }}
      </span>
      <span class="calendar-event bg-warning text-warning text-center" v-if="localAttendance.attending_status === 'attend_late'">
        {{ $t(`meta.attendance_statuses.${localAttendance.attending_status}`) }}
      </span>
      <span href="#" class="calendar-event bg-success text-success text-center" v-if="localAttendance.leaving_status === 'leave_ok'">
        {{ $t(`meta.attendance_statuses.${localAttendance.leaving_status}`) }}
      </span>
      <span href="#" class="calendar-event bg-error text-error text-center" v-if="localAttendance.leaving_status === 'leave_early'">
        {{ $t(`meta.attendance_statuses.${localAttendance.leaving_status}`) }}
      </span>
      <span href="#" class="calendar-event bg-info text-info text-center" v-if="localAttendance.off_status">
        {{ $t(`meta.attendance_statuses.${localAttendance.off_status}`) }}
      </span>
      <span href="#" class="calendar-event bg-gray text-gray text-center" v-if="localAttendance.holiday">
        {{ localAttendance.holiday.name }}
      </span>
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex'

export default {
  name: 'calendar-date',

  data () {
    return {
      localAttendance: this.calendarAttendance
    }
  },

  props: ['calendarAttendance', 'today'],

  computed: {
    ...mapState('punch', [
      'attendance'
    ]),

    ...mapState('initialStates', [
      'currentCompany',
      'meta'
    ]),

    isWeekend () {
      return this.currentCompany.breakdays.includes(this.$moment(this.localAttendance.day).format('dddd').toLowerCase())
    }
  },

  watch: {
    attendance: {
      handler: function (newAttendance) {
        if (newAttendance.day === this.localAttendance.day) this.localAttendance = newAttendance
      },
      deep: true
    }
  }
}
</script>
