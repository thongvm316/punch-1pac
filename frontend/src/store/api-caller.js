import axios from 'axios'

export default function callApi(payload = { params: null, data: null }) {
  const { method, url, params, data } = payload

  return axios({
    method,
    url,
    params,
    data
  })
}
