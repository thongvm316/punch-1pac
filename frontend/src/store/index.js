import Vue from 'vue'
import Vuex from 'vuex'
import createLogger from 'vuex/dist/logger'

import headerAnnouncements from './modules/header-announcements'
import punch from './modules/punch'

import companyUsers from './modules/company-users'
import companyDepartments from './modules/company-departments'

import announcements from './modules/announcements'
import announcement from './modules/announcement'
import userSessions from './modules/user-sessions'
import attendances from './modules/attendances'
import companyAllowedIPs from './modules/company-allowed_ips'

Vue.use(Vuex)

const debug = process.env.NODE_ENV !== 'production'

export default new Vuex.Store({
  modules: {
    headerAnnouncements,
    punch,
    companyUsers,
    companyDepartments,
    announcements,
    announcement,
    userSessions,
    attendances,
    companyAllowedIPs
  },
  strict: debug,
  plugins: debug ? [createLogger()] : []
})
