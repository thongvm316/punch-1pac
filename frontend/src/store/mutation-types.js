// initial-states
export const INITIAL_STATES_SET_USER = 'INITIAL_STATES_SET_USER'
export const INITIAL_STATES_SET_COMPANY = 'INITIAL_STATES_SET_COMPANY'
export const INITIAL_STATES_SET_META = 'INITIAL_STATES_SET_META'
export const INITIAL_STATES_UPDATE_USER = 'INITIAL_STATES_UPDATE_USER'
export const INITIAL_STATES_UPDATE_COMPANY = 'INITIAL_STATES_UPDATE_COMPANY'
export const INITIAL_STATES_SET_USER_ERRORS = 'INITIAL_STATES_SET_USER_ERRORS'
export const INITIAL_STATES_SET_COMPANY_ERRORS = 'INITIAL_STATES_SET_COMPANY_ERRORS'
export const INITIAL_STATES_CLEAR_USER_ERRORS = 'INITIAL_STATES_CLEAR_USER_ERRORS'
export const INITIAL_STATES_CLEAR_COMPANY_ERRORS = 'INITIAL_STATES_CLEAR_COMPANY_ERRORS'
export const INITIAL_STATES_UPDATE_PASSWORD_CHANGED = 'INITIAL_STATES_UPDATE_PASSWORD_CHANGED'

// announcements
export const READ_ANNOUNCEMENT = 'READ_ANNOUNCEMENT'
export const RECEIVE_HEADER_ANNOUNCEMENTS = 'RECEIVE_HEADER_ANNOUNCEMENTS'

// attendances
export const RECEIVE_ATTENDANCES = 'RECEIVE_ATTENDANCES'

// punch
export const PUNCH_IN = 'PUNCH_IN'
export const PUNCH_OUT = 'PUNCH_OUT'
export const PUNCH_INIT_ATTENDANCE = 'PUNCH_INIT_ATTENDANCE'

// company-users
export const FETCH_USERS = 'FETCH_USERS'
export const DELETE_USER = 'DELETE_USER'
export const UPDATE_USER = 'UPDATE_USER'

// company-business-days
export const FETCH_BUSINESS_DAYS = 'FETCH_BUSINESS_DAYS'
export const ADD_BUSINESS_DAY = 'ADD_BUSINESS_DAY'
export const DELETE_BUSINESS_DAY = 'DELETE_BUSINESS_DAY'
export const UPDATE_BUSINESS_DAY = 'UPDATE_BUSINESS_DAY'
export const UPDATE_BUSINESS_DAY_ERRORS = 'UPDATE_BUSINESS_DAY_ERRORS'
export const CLEAR_BUSINESS_DAY_ERRORS = 'CLEAR_BUSINESS_DAY_ERRORS'

// company-allowed_ips
export const FETCH_IPS = 'FETCH_IPS'
export const CREATE_IP = 'CREATE_IPS'
export const UPDATE_IP = 'UPDATE_IPS'
export const DELETE_IP = 'DELETE_IPS'
export const UPDATE_IP_ERRORS = 'UPDATE_REQUEST_ERRORS'
export const CLEAR_IP_ERRORS = 'CLEAR_REQUEST_ERRORS'

// company-holidays
export const FETCH_HOLIDAYS = 'FETCH_HOLIDAYS'
export const CREATE_HOLIDAY = 'CREATE_HOLIDAY'
export const UPDATE_HOLIDAY = 'UPDATE_HOLIDAY'
export const DELETE_HOLIDAY = 'DELETE_HOLIDAY'
export const UPDATE_HOLIDAY_ERRORS = 'UPDATE_HOLIDAY_ERRORS'
export const CLEAR_HOLIDAY_ERRORS = 'CLEAR_HOLIDAY_ERRORS'
export const IMPORT_NATIONAL_HOLIDAYS = 'IMPORT_NATIONAL_HOLIDAYS'

// user-sessions
export const FETCH_SESSIONS = 'FETCH_SESSIONS'
export const DELETE_SESSION = 'DELETE_SESSION'

// user-password
export const UPDATE_USER_PASSWORD_ERRORS = 'UPDATE_USER_PASSWORD_ERRORS'
export const CLEAR_USER_PASSWORD_ERRORS = 'CLEAR_USER_PASSWORD_ERRORS'

// requests
export const RECEIVE_REQUESTS = 'RECEIVE_REQUESTS'
export const ADD_REQUEST = 'ADD_REQUEST'
export const UPDATE_REQUEST = 'UPDATE_REQUEST'
export const DELETE_REQUEST = 'DELETE_REQUEST'
export const UPDATE_REQUEST_ERRORS = 'UPDATE_REQUEST_ERRORS'
export const CLEAR_REQUEST_ERRORS = 'CLEAR_REQUEST_ERRORS'

// group-requests
export const RECEIVE_GROUP_REQUESTS = 'RECEIVE_GROUP_REQUESTS'
export const APPROVE_GROUP_REQUEST = 'APPROVE_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST = 'REJECT_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST_ERRORS = 'REJECT_GROUP_REQUEST_ERRORS'
export const CLEAR_REJECT_GROUP_REQUEST_ERRORS = 'CLEAR_REJECT_GROUP_REQUEST_ERRORS'

// group-attendances
export const RECEIVE_GROUP_ATTENDANCES = 'RECEIVE_GROUP_ATTENDANCES'
export const FETCH_USERS_IN_GROUP_ATTENDANCES = 'FETCH_USERS_IN_GROUP_ATTENDANCES'

// group-report
export const FETCH_GROUP_REPORT = 'FETCH_GROUP_REPORT'

// groups
export const RECEIVE_GROUPS = 'RECEIVE_GROUPS'
export const ADD_GROUP = 'ADD_GROUP'
export const UPDATE_GROUPS_ERRORS = 'UPDATE_GROUPS_ERRORS'
export const CLEAR_GROUPS_ERRORS = 'CLEAR_GROUPS_ERRORS'

// group
export const RECEIVE_GROUP = 'RECEIVE_GROUP'
export const UPDATE_GROUP = 'UPDATE_GROUP'
export const ADD_GROUP_USER = 'ADD_GROUP_USER'
export const REMOVE_GROUP_USER = 'REMOVE_GROUP_USER'
export const UPDATE_GROUP_ERRORS = 'UPDATE_GROUP_ERRORS'
export const CLEAR_GROUP_ERRORS = 'CLEAR_GROUP_ERRORS'
export const FETCH_USERS_NOT_IN_GROUP = 'FETCH_USERS_NOT_IN_GROUP'

// chart
export const FETCH_CHART_DATA = 'FETCH_CHART_DATA'
export const RESET_CHART_DATA = 'RESET_CHART_DATA'

// calendar
export const FETCH_CALENDAR_ATTENDANCES = 'FETCH_CALENDAR_ATTENDANCES'

// flash
export const SET_FLASH_MESSAGE = 'SET_FLASH_MESSAGE'

// notifications
export const FETCH_HEADER_NOTIFICATIONS = 'FETCH_HEADER_NOTIFICATIONS'
export const FETCH_NEW_HEADER_NOTIFICATIONS = 'FETCH_NEW_HEADER_NOTIFICATIONS'
export const FETCH_MORE_HEADER_NOTIFICATIONS = 'FETCH_MORE_HEADER_NOTIFICATIONS'
export const READ_NOTIFICATIONS = 'READ_NOTIFICATIONS'
export const SET_INTERVAL_FETCH_NOTIFICATIONS = 'SET_INTERVAL_FETCH_NOTIFICATIONS'

// activities
export const FETCH_ACTIVITIES = 'FETCH_ACTIVITIES'
export const FETCH_MORE_ACTIVITIES = 'FETCH_MORE_ACTIVITIES'

// status-cards
export const FETCH_STATUS_CARDS = 'FETCH_STATUS_CARDS'
