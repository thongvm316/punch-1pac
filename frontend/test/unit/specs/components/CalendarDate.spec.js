import { shallowMount } from '@vue/test-utils'

import localVue from '../../supports/local-vue'
import setComputed from '../../supports/set-computed'

import store from '@/store'
import i18n from '@/locale'
import CalendarDate from '@/components/CalendarDate'

const propsData = {
  normalDay: {
    calendarAttendance: {
      attended_at: '08:13',
      attending_status: 'attend_ok',
      day: localVue.prototype.$moment().format('YYYY-MM-DD'),
      leaving_status: 'leave_early',
      left_at: '08:13'
    },

    today: localVue.prototype.$moment()
  },

  offDay: {
    calendarAttendance: {
      day: '2019-12-03',
      off_status: 'annual_leave'
    },

    today: localVue.prototype.$moment()
  },

  weekendDay: {
    calendarAttendance: {
      day: '2019-12-08',
    },

    today: localVue.prototype.$moment()
  },

  specialDay: {
    calendarAttendance: {
      day: '2019-12-13',
      holiday: {
        name: 'Tet holiday'
      }
    },

    today: localVue.prototype.$moment()
  }
}

describe('CalendarDate.vue', () => {
  let wrapper

  afterEach(() => { wrapper.vm.$destroy() })

  // Normal date
  describe('Normal date', () => {
    beforeEach(() => {
      wrapper = shallowMount(CalendarDate, {
        i18n,
        store,
        propsData: propsData.normalDay,
        localVue
      })
    })

    describe('when CalendarDate was mounted', () => {
      it('should render correctly', () => {
        expect(wrapper.exists()).toBe(true)
        expect(wrapper.isVueInstance()).toBeTruthy()
      })
    })

    describe('when validation propsData', () => {
      it('should validate props', () => {
        const { calendarAttendance, today } = wrapper.vm.$options.props

        expect(calendarAttendance.required).toBeTruthy
        expect(calendarAttendance.type).toBe(Object)
        expect(today.required).toBeTruthy
        expect(today.type).toBe(Object)
      })
    })

    describe('when tooltip && calendar event', () => {
      let calendarDateEl
      let calendarEventEls

      beforeEach(() => {
        calendarDateEl = wrapper.find('.calendar-date')
        calendarEventEls = wrapper.findAll('.calendar-event')
      })

      it('should display data-tooltip with attending_status && leaving_status', () => {
        expect(calendarDateEl.classes()).toContain('tooltip')
        expect(calendarDateEl.attributes('data-tooltip')).toEqual('In: 08:13 - Out: 08:13')
        expect(calendarDateEl.classes()).not.toContain('disabled')
      })

      it('should display calendar event', () => {
        expect(calendarEventEls).toHaveLength(2)
        expect(calendarEventEls.at(0).classes()).toContain('text-success')
        expect(calendarEventEls.at(0).text()).toEqual('Attend OK')
        expect(calendarEventEls.at(1).classes()).toContain('text-error')
        expect(calendarEventEls.at(1).text()).toEqual('Leave Early')
      })

      it('should display today', () => {
        expect(wrapper.find('button.date-item').classes()).toContain('date-today')
        expect(wrapper.find('button.date-item').text()).toEqual(wrapper.vm.localAttendance.day.split('-')[2])
      })
    })

    describe('when watcher attendance', () => {
      let calendarDateEl
      let calendarEventEls

      beforeEach(() => {
        calendarDateEl = wrapper.find('.calendar-date')
        calendarEventEls = wrapper.findAll('.calendar-event')
      })

      it('should render new attendance', () => {
        setComputed(wrapper, { attendance: {
          attended_at: '08:13',
          attending_status: 'attend_late',
          day: localVue.prototype.$moment().format('YYYY-MM-DD'),
          leaving_status: 'leave_ok',
          left_at: '10:13'
        } })

        expect(calendarEventEls).toHaveLength(2)
        expect(calendarEventEls.at(0).classes()).toContain('text-warning')
        expect(calendarEventEls.at(0).text()).toEqual('Attend Late')
        expect(calendarEventEls.at(1).classes()).toContain('text-success')
        expect(calendarEventEls.at(1).text()).toEqual('Leave OK')
      })
    })
  })

  // Off date
  describe('when off date', () => {
    beforeEach(() => {
      wrapper = shallowMount(CalendarDate, {
        i18n,
        store,
        propsData: propsData.offDay,
        computed: {
          currentCompany: () => {
            return { breakdays: ['saturday', 'sunday'] }
          }
        },

        localVue
      })
    })

    describe('when tooltip', () => {
      let calendarDateEl
      let calendarEventEls

      beforeEach(() => {
        calendarDateEl = wrapper.find('.calendar-date')
        calendarEventEls = wrapper.findAll('.calendar-event')
      })

      it('should no tooltip', () => {
        expect(calendarDateEl.classes()).not.toContain('tooltip')
        expect(calendarDateEl.attributes('data-tooltip')).toEqual('')
      })

      it('should display calendar event', () => {
        expect(calendarEventEls).toHaveLength(1)
        expect(calendarEventEls.at(0).text()).toEqual('Day Off')
      })
    })
  })

  // Special day
  describe('Special day', () => {
    beforeEach(() => {
      wrapper = shallowMount(CalendarDate, {
        i18n,
        store,
        propsData: propsData.specialDay,
        computed: {
          currentCompany: () => {
            return { breakdays: ['saturday', 'sunday'] }
          }
        },
        localVue
      })
    })

    describe('when tooltip', () => {
      let calendarDateEl
      let calendarEventEls

      beforeEach(() => {
        calendarDateEl = wrapper.find('.calendar-date')
        calendarEventEls = wrapper.findAll('.calendar-event')
      })

      it('should display holiday tooltip', () => {
        expect(calendarDateEl.classes()).toContain('tooltip')
        expect(calendarDateEl.attributes('data-tooltip')).toEqual('Tet holiday')
      })

      it('should display calendar event', () => {
        expect(calendarEventEls).toHaveLength(1)
        expect(calendarEventEls.at(0).text()).toEqual('Tet holiday')
      })
    })
  })

  // Weekend day
  describe('Weekend day', () => {
    beforeEach(() => {
      wrapper = shallowMount(CalendarDate, {
        i18n,
        store,
        propsData: propsData.weekendDay,
        computed: {
          currentCompany: () => {
            return { breakdays: ['saturday', 'sunday'] }
          }
        },

        localVue
      })
    })

    describe('when tooltip', () => {
      let calendarDateEl
      let calendarEventEls

      beforeEach(() => {
        calendarDateEl = wrapper.find('.calendar-date')
        calendarEventEls = wrapper.findAll('.calendar-event')
      })

      it('should display holiday tooltip', () => {
        expect(calendarDateEl.classes()).not.toContain('tooltip')
        expect(calendarDateEl.classes()).toContain('disabled')
        expect(calendarDateEl.attributes('data-tooltip')).toEqual('')
      })

      it('should display calendar event', () => {
        expect(calendarEventEls).toHaveLength(0)
      })
    })
  })
})
