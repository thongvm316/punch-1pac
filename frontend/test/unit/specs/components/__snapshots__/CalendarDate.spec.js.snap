// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarDate.vue when normal date when CalendarDate was mounted should render correctly 1`] = `
<div data-tooltip="In: 09:45 - Out: 09:45" class="calendar-date current-month tooltip"><button class="date-item">
    01
  </button>
  <div class="calendar-events"><span class="calendar-event bg-success text-success text-center">
      Attend OK
    </span>
    <!----> <span href="#" class="calendar-event bg-success text-success text-center">
      Leave OK
    </span>
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;

exports[`CalendarDate.vue when normal date when watcher attendance should render new attendance 1`] = `
<div data-tooltip="In: 10:45 - Out: 20:45" class="calendar-date current-month tooltip"><button class="date-item">
    01
  </button>
  <div class="calendar-events">
    <!----> <span class="calendar-event bg-warning text-warning text-center">
      Attend Late
    </span> <span href="#" class="calendar-event bg-success text-success text-center">
      Leave OK
    </span>
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;

exports[`CalendarDate.vue when off date when CalendarDate was mounted should render correctly 1`] = `
<div data-tooltip="" class="calendar-date current-month"><button class="date-item">
    03
  </button>
  <div class="calendar-events">
    <!---->
    <!---->
    <!---->
    <!----> <span href="#" class="calendar-event bg-notice text-notice text-center">
      Day Off
    </span>
    <!---->
  </div>
</div>
`;

exports[`CalendarDate.vue when special day when CalendarDate was mounted should render correctly 1`] = `
<div data-tooltip="Tet holiday" class="calendar-date current-month disabled tooltip"><button class="date-item">
    13
  </button>
  <div class="calendar-events">
    <!---->
    <!---->
    <!---->
    <!---->
    <!----> <span href="#" class="calendar-event bg-gray text-gray text-center">
      Tet holiday
    </span></div>
</div>
`;

exports[`CalendarDate.vue when week day when CalendarDate was mounted should render correctly 1`] = `
<div data-tooltip="" class="calendar-date current-month disabled"><button class="date-item">
    08
  </button>
  <div class="calendar-events">
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;
